identity/src/main/java/com/android/identity/CborUtil.java

```diff
@@ -0,0 +1,305 @@
+import java.text.DecimalFormatSymbols;
+import java.util.Collection;
+import java.util.List;
+import java.util.Locale;
+
+import co.nstant.in.cbor.CborDecoder;
+import co.nstant.in.cbor.CborException;
+import co.nstant.in.cbor.model.AbstractFloat;
+import co.nstant.in.cbor.model.ByteString;
+import co.nstant.in.cbor.model.DataItem;
+import co.nstant.in.cbor.model.DoublePrecisionFloat;
+import co.nstant.in.cbor.model.NegativeInteger;
+import co.nstant.in.cbor.model.SimpleValue;
+import co.nstant.in.cbor.model.UnsignedInteger;
+
+class CborUtil {
+    private static final String TAG = "CborUtil";
+
+
+    static final int DIAGNOSTICS_FLAG_EMBEDDED_CBOR = (1 << 0);
+
+    static final int DIAGNOSTICS_FLAG_PRETTY_PRINT = (1 << 1);
+
+    @Retention(SOURCE)
+    @IntDef(
+            flag = true,
+            value = {
+                    DIAGNOSTICS_FLAG_EMBEDDED_CBOR,
+                    DIAGNOSTICS_FLAG_PRETTY_PRINT
+            })
+    public @interface DiagnosticsFlags {
+    }
+
+    // Returns true iff all elements in |items| are not compound (e.g. an array or a map).
+    private static boolean cborAreAllDataItemsNonCompound(@NonNull List<DataItem> items,
+                                                          @DiagnosticsFlags int flags) {
+        for (DataItem item : items) {
+            if ((flags & CborUtil.DIAGNOSTICS_FLAG_EMBEDDED_CBOR) != 0 && item.hasTag()) {
+                return false;
+            }
+            switch (item.getMajorType()) {
+                case ARRAY:
+                case MAP:
+                    return false;
+
+                default:
+                    // Do nothing
+                    break;
+            }
+        }
+        return true;
+    }
+
+    private static boolean cborFitsInSingleLine(@NonNull List<DataItem> items,
+                                                @DiagnosticsFlags int flags) {
+        return cborAreAllDataItemsNonCompound(items, flags) && items.size() < 8;
+    }
+
+    private static void toDiagnostics(@NonNull StringBuilder sb,
+                                      int indent,
+                                      @NonNull DataItem dataItem,
+                                      @DiagnosticsFlags int flags) {
+        int count;
+
+        boolean pretty = ((flags & DIAGNOSTICS_FLAG_PRETTY_PRINT) != 0);
+        String indentString = "";
+        if (pretty) {
+            StringBuilder indentBuilder = new StringBuilder();
+            for (int n = 0; n < indent; n++) {
+                indentBuilder.append(' ');
+            }
+            indentString = indentBuilder.toString();
+        }
+
+        if (dataItem.hasTag()) {
+            sb.append(String.format(Locale.US, "%d(", dataItem.getTag().getValue()));
+        }
+
+        switch (dataItem.getMajorType()) {
+            case INVALID:
+                sb.append("<invalid>");
+                break;
+
+            case UNSIGNED_INTEGER:
+                // Major type 0: an unsigned integer.
```

Comment:
Uber nit:  I don't think the comments at the top of each case add anything at all. The case label already explains it, especially when considered in context of the `switch (dataItem.getMajorType())`.

This triggers my "comments are a code smell" interrupt... though usually the code smell is that the comments are necessary because the code could/should be clearer. In this case, the code is clear without them.

Reply:
This tells you that this is Major Type 0 which I think is actually helpful.